### PR TITLE
Add new IntPtr/UIntPtr API surface

### PR DESCRIPTION
**Pull request by john-h-k**
Implements https://github.com/dotnet/corefx/issues/20256

Delegates all implementations to `[U]Int32/[U]Int64`. Uses `InvariantCulture` for `ToString` to maintain backcompat for the old `ToString()` method

### CODE REVIEW

**Comment by tannergooding** [comment]
CC. @cston, @333fred

File: src/libraries/System.Private.CoreLib/src/System/IntPtr.cs
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/IntPtr.cs]: Why is this needed? structs are Sequential by default.

File: src/libraries/System.Private.CoreLib/src/System/IntPtr.cs
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/IntPtr.cs]: nit: I think the conventionis to do `((nint)_value)` rather than `((nint)this)`

File: src/libraries/System.Private.CoreLib/src/System/IntPtr.cs
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/IntPtr.cs]: The Min/MaxValue properties should be marked with `[NonVersionable]` attribute.

File: src/libraries/System.Private.CoreLib/src/System/IntPtr.cs
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/IntPtr.cs]: Mark Equals with `[NonVersionable]` attribute

**Comment by tannergooding** [comment]
The reference assembly also needs to be updated and tests added, but I don't believe we can do it in the same PR until after https://github.com/dotnet/runtime/issues/92 is merged.

File: src/libraries/System.Private.CoreLib/src/System/IntPtr.cs
Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/IntPtr.cs]: We also have it on the other primitive types, such as `Int32`. It was added for consistency.

If we don't add it here, we should likely have a separate PR that removes it elsewhere.

File: src/libraries/System.Private.CoreLib/src/System/IntPtr.cs
Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/IntPtr.cs]: Do both forms generate same IL? We should make sure that the native code generated for these is good.

File: src/libraries/System.Private.CoreLib/src/System/IntPtr.cs
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/IntPtr.cs]: There are number of places throughout CoreLib that assume IntPtr/UIntPtr are not comparable. We should fix them up as necessary.

For example, https://github.com/dotnet/runtime/blob/4f9ae42d861fcb4be2fcd5d3d55d5f227d30e723/src/libraries/System.Private.CoreLib/src/System/Array.cs#L513 should be fixed up.

File: src/libraries/System.Private.CoreLib/src/System/IntPtr.cs
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/IntPtr.cs]: Maybe use `Unsafe.As` to avoid the copy?

File: src/libraries/System.Private.CoreLib/src/System/IntPtr.cs
Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/IntPtr.cs]: @john-h-k, could you get an assembly diff for the two forms?

I would hope the JIT folds them the same, but the latter ends up going through the implicit conversion API (vs directly converting the field).

We were already using the former for existing APIs (like `ToString()`).

File: src/libraries/System.Private.CoreLib/src/System/UIntPtr.cs
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/UIntPtr.cs]: The same feedback should be applied in this file as well.

Comment (john-h-k) [reply src/libraries/System.Private.CoreLib/src/System/UIntPtr.cs]: believe ive resolved them all

**Comment by danmosemsft** [comment]
@tannergooding what is next here? Are you waiting on the assembly diff mentioned above? Or is this ready for re-review?

**Comment by jkotas** [comment]
This PR needs:
- Matching change in reference assemblies
- Tests
- Addressing all feedback above (e.g. https://github.com/dotnet/runtime/pull/307#discussion_r350938305 )

**Comment by tannergooding** [comment]
It is waiting on the assembly diffs and the remaining feedback from Jan to be addressed. @john-h-k, please let us know if you need any assistance in getting this completed or if you have any questions.

**Comment by tannergooding** [comment]
Thanks for rebasing this @john-h-k.

You'll need to update the reference assembly: https://github.com/dotnet/runtime/blob/master/src/libraries/System.Runtime/ref/System.Runtime.cs#L2391
You should be able to accomplish this by following: https://github.com/dotnet/runtime/blob/master/docs/coding-guidelines/updating-ref-source.md

We also need the following two completed before merging can happen:
* Tests
* Addressing all feedback above (e.g. #307 (comment) )

Please let us know if you need any assistance!

File: src/libraries/System.Private.CoreLib/src/System/IntPtr.cs
Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/IntPtr.cs:24]: @jkotas, do we have a rough list of these places?

It looks like 2 places in System.Array and no others for `ELEMENT_TYPE_I` or `ELEMENT_TYPE_U` (in managed code, I didn't check native):
* https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/Array.cs#L515
* https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/Array.cs#L1694

I did a quick scan of some other places and didn't spot anything that was keying off `typeof(IntPtr` or `typeof(UIntPtr)`. Most places look to just key off `x is IComparable`.

**Comment by danmosemsft** [comment]
New surface needs new API docs. @carlossanlop should we have /// comments in here? Or simply open a separate PR against https://github.com/dotnet/dotnet-api-docs/blob/master/xml/System/UIntPtr.xml (how do we avoid losing track?)

**Comment by carlossanlop** [comment]
> New surface needs new API docs. @carlossanlop should we have /// comments in here?

Thanks @danmosemsft. I added the label.

@john-h-k can you please make sure to decorate the new public API implementations in the src file with triple slash documentation? It's a requirement to approve the PR for merging.

@maryamariyan do you know why the bot did not add the label automatically for this PR?

**Comment by maryamariyan** [comment]
@carlossanlop it would have added that label had it been created today.

File: src/libraries/System.Private.CoreLib/src/System/Array.cs
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Array.cs:1463]: It was done this way to reduce number of generic instantations that this needs. This is undoing that optimization.

File: src/libraries/System.Runtime/ref/System.Runtime.cs
Comment (tannergooding) [inline src/libraries/System.Runtime/ref/System.Runtime.cs:4367]: I just noticed that `ShiftMask` is missing. These should be exposed as part of the PR and should be properties as per: https://github.com/dotnet/runtime/issues/21943#issuecomment-555895614

Comment (cston) [reply src/libraries/System.Runtime/ref/System.Runtime.cs:4367]: The C# compiler will emit the equivalent code, `(sizeof(nint) * 8) - 1`, explicitly rather than using the `ShiftMask` members, so the members are not needed for the compiler.

---
In reply to: [408345037](https://github.com/dotnet/runtime/pull/307#discussion_r408345037) [](ancestors = 408345037)

Comment (tannergooding) [reply src/libraries/System.Runtime/ref/System.Runtime.cs:4367]: Thanks @cston. We shouldn't need to expose that member if the compiler won't actually use it. CC. @terrajobst, @dotnet/fxdc

File: src/libraries/System.Private.CoreLib/src/System/Array.cs
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/Array.cs:1273]: There is some dead code here, likely just an issue in undoing the previous change

File: src/libraries/System.Private.CoreLib/src/System/Array.cs
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/Array.cs:1463]: It would be nice to minimize this diff by putting `R4` back where it was originally (same with `R8` below).

File: src/libraries/System.Private.CoreLib/src/System/Array.cs
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/Array.cs:515]: I think the handling for `ELEMENT_TYPE_I` and `ELEMENT_TYPE_U` is missing just below.

We likely want to have a test covering these methods as well.

File: src/libraries/System.Private.CoreLib/src/System/Array.cs
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/Array.cs:1697]: I think we can use the same trick as was done for `Reverse` here, that is treat `ELEMENT_TYPE_I` as `I4` on 32-bit and as `I8` on 64-bit, so we can avoid the additional generic instantiation.

File: src/libraries/System.Private.CoreLib/src/System/Array.cs
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/Array.cs:513]: Missing a paren here.

File: src/libraries/System.Private.CoreLib/src/Resources/Strings.resx
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/Resources/Strings.resx:569]:
```suggestion
    <value>Object must be of type UIntPtr.</value>
```

**Review by tannergooding — APPROVED** [review]
LGTM. Thanks for working through this and getting it done @john-h-k

**Comment by tannergooding** [comment]
@jkotas, would you be able to give this a quick review? It would be nice to get this in before the snap today so it can make the same preview as the compiler support for `nint`/`nuint`.

**Review by jkotas — APPROVED** [review]
LGTM

File: src/libraries/System.Private.CoreLib/src/System/IntPtr.cs
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/IntPtr.cs:207]: Test failures are because this is forcing `InvariantCulture`.

Either this needs to just mirror what `int32`/`int64` do for their respective overloads or the test needs to be updated. I would guess the former is preferred and so this should just be `((nint)_value).ToString()`

Comment (john-h-k) [reply src/libraries/System.Private.CoreLib/src/System/IntPtr.cs:207]: that is a breaking change as previously it was invariant by default right?

Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/IntPtr.cs:207]: Yes, but it is also a breaking change to implement `IFormattable` (hence the `breaking-change` label on https://github.com/dotnet/runtime/issues/21943).

I think it is the least confusing to just make this consistent moving forward and we can direct users to explicitly specify `CultureInfo.InvariantCulture` moving forward, if they want to be backwards compatible. Otherwise, you end up with behavior that differs from the other primitive types and how these overloads generally work together in general.

(Unless @jkotas or someone else has a different opinion here).

Comment (john-h-k) [reply src/libraries/System.Private.CoreLib/src/System/IntPtr.cs]: changed as this is already a breaking change, discussed w tanner somewhere else

**Pull request merged**
